syntax="proto2";
package elasticfaiss;
option cc_generic_services = true;

import "shard.proto";

enum WorkNodeState{
    WNODE_ACTIVE  = 1; 
    WNODE_TIMEOUT = 2;
}

message WorkNode{
    required string peer_id = 1;
    optional int64 last_active_ms = 2;
    optional int64 last_state_ms  = 3;
    repeated Shard shards = 4;
    required int32 state = 5;
}

message ClusterState{
    optional string name = 1;
    optional int64  update_ms = 2;
    repeated WorkNode nodes = 3; 
}

message MasterSnapshot{
    repeated  ClusterState states = 1;
}

message BootstrapRequest {
    required string cluster   = 1;
    required string node_peer = 2;
    optional int64  boot_ms   = 3;
};

message BootstrapResponse {
    required bool success = 1;
    optional string redirect = 2;
    repeated WorkNode nodes = 3;    
};

message NodeHeartbeatRequest {
    required string cluster    = 1;
    required string node_peer  = 2;
    optional int64  active_ms  = 3;
    repeated Shard  shards     = 4;
};

message NodeHeartbeatResponse {
    required bool success = 1;
    optional string redirect = 2;
    repeated WorkNode nodes = 3; 
};

message GetClusterStateRequest{
    required string cluster   = 1;
};

message GetClusterStateResponse{
    required bool success = 1;
    optional string redirect = 2;
    optional ClusterState state   = 3;
};



message CreateIndexRequest {
    required  string cluster    = 1;
    required  IndexConf conf = 2;
};

message CreateIndexResponse {
    required bool success = 1;
    optional string redirect = 2;
    optional string error = 3;
};

message DeleteIndexRequest {
    required  string cluster    = 1;
    required string index_name = 2;
};

message DeleteIndexResponse {
    required bool success = 1;
    optional string redirect = 2;
    optional string error = 3;
};

message UpdateIndexRequest {
    required  string cluster    = 1;
    required  IndexConf conf    = 2;
};

message UpdateIndexResponse {
    required bool success = 1;
    optional string redirect = 2;
    optional string error = 3;
};


service MasterService {
    rpc bootstrap(BootstrapRequest) returns (BootstrapResponse);
    rpc node_heartbeat(NodeHeartbeatRequest) returns (NodeHeartbeatResponse);
    rpc get_cluster_state(GetClusterStateRequest) returns (GetClusterStateResponse);
    
    rpc create_index(CreateIndexRequest) returns (CreateIndexResponse);
    rpc delete_index(DeleteIndexRequest) returns (DeleteIndexResponse); 
    rpc update_index(UpdateIndexRequest) returns (UpdateIndexResponse); 
};


